Parser for the Agriculture Sensor

flixbi opened the pull request:
@h4cc wenn ich darf erweitere ich hiermit ein wenig eure Parser-Bibliothek ^^
=> Rechtlich dürfte das kein Problem sein, da ich der Author bin und der Parser durch einen Gefallen entstand (und keinen Auftrag)
✊ für die OpenSource Community

h4cc committed:
Added comtac_klax.ex parser

Commit:
Merge branch 'master' into smilio_action

Commit:
Add Smilio Action parser to readme

Commit:
Add fields for buttons

Commit:
Add comment to describe parser

Commit:
improve documentation

Commit:
replace unsupported guard by splitting it up into seperate functions
+ add message_type to data frame

Commit:
add message_type to  fields

h4cc committed:
Added option add_power_from_last_reading? to dzg v2 parser.

h4cc committed:
Added bosch parksensor.

h4cc committed:
Added bosch parksensor to readme.

h4cc committed:
Added bosch parksensor link to PDF.

h4cc committed:
Fixed DZG Parser for restrictions of ELEMENT.

h4cc committed:
Elsys v8: Added sTypes 03, 0F, 14. Fields and  Tests.

h4cc committed:
Elsys v8: Added acceleration_sum

Mr-SvomF committed:
Create example_ieee754

h4cc committed:
comtac_klax: 2019-03-04 [jb]: Skipping invalid backdated values when value==0.0

Mr-SvomF committed:
Rename example_ieee754 to example_ieee754.ex

Mr-SvomF committed:
Update README.md

h4cc committed:
comtac_klax: 2019-03-04 [jb]: Skipping invalid backdated values when value==0.0; Added mode "Logarex"

h4cc committed:
Added NKE Watteco IN'O in README.

h4cc committed:
Added NKE Watteco IN'O in README.

h4cc committed:
comtac_klax: 2019-03-05 [jb]: Register values now signed. Added fields().

h4cc committed:
Fixed "Humidity Sensor" for real payload.

h4cc committed:
Merge pull request #12 from ZennerIoT/smilio_action

Parser for Smilio Action FW 2.0.1.x

h4cc committed:
2019-03-27 [jb]: Logging unknown binaries in _parse_payload/2.

h4cc committed:
Testing if all parsers are linked in README.

h4cc committed:
2019-04-04 [jb]: Fixed invalid "temperature" value and typo. Added tests.

h4cc committed:
2019-04-04 [jb]: Skipping 65535 for co2 and voc. Added tests.

h4cc committed:
2019-04-04 [jb]: Skipping 65535 for co2 and voc. Added tests. Added parse() fallback.

h4cc committed:
2019-04-04 [jb]: Skipping 65535 for co2 and voc. Added tests. Added parse() fallback.

Commit:
tabs_doornwindow: 2019-04-04 [gw]: Adding tests(), formatted code, corrected name of device in comment.

Commit:
tabs_motion: 2019-04-04 [gw]: Corrected name of device in comment.

Commit:
tabs_object-locator: 2019-04-04 [gw]: Added fields(), added frameport to tests.

Commit:
tabs_pushbutton: 2019-04-04 [gw]: formatted code, corrected name of device in comment, added frameport to tests

Commit:
Initial version of tabs.ex

Commit:
tabs_doornwindow: use public instead of private functions

Commit:
tabs:
 * improve map creation
 * add fields for states as number
 * refactor
 * adjust tests to work with the new fields
 * added new fields to fields/0

Commit:
tabs:
 * added comments to old tabs parsers and README

kleinernik committed:
add state field back for map visualisation

kleinernik committed:
Update tabs_doornwindow.ex

Commit:
dzg_loramod: Handle medium electricity with qualifier A_Plus

Commit:
Merge branch 'dzg'

Commit:
bosch_parksensor: change order of bytes in startup message according to FW v0.23.3

gerritwalther committed:
Merge pull request #16 from ZennerIoT/bosch

bosch_parksensor: change order of bytes in startup message according …

Commit:
dzg_loramod: use String.to_atom/1 instead of :erlang.binary_to_atom/1

Commit:
dzg_loramod: return multiple readings when receiving values from a A_Plus qualifier.

Commit:
nas_um3023:
* Updated according to new information from 0.7.0 docs
* fixed rssi value
* fixed medium type mapping
* correctly implemented analog value parsing

Commit:
nas_um3023:
* renamed file to correctly reflect the device

Commit:
nas_um3023:
* remove needless test

Mr-SvomF committed:
Create nas_luminaire.ex

initial commit

Mr-SvomF committed:
Update README.md

added nas_luminaire.ex

Commit:
nas_um30x3:
* handling nas um3023 and 3033 in one parser
* using internal wmbus lib to parse mbus data from nas 3033 device
* wmbus lib can be used in :test environment
** uncomment test to validate functionality
* lots of refactoring

kleinernik committed:
Create rak_button.ex

h4cc committed:
Added lib/example_template.ex

Commit:
nas_um3033:
* now handling mbus connect message on fport 53

h4cc committed:
Merge pull request #17 from kleinernik/patch-5

Create rak_button.ex

h4cc committed:
Fixed parser for RAK button.

h4cc committed:
Added RAK Button to readme

Commit:
zis_smart-waste-ultrasonic:
* Refactor

Commit:
bosch_parksensor: remove warning

h4cc committed:
Using lib_wmbus in every build using travis.

h4cc committed:
Testing builds using travis with internal libraries.

h4cc committed:
Testing builds using travis with internal libraries. Fix1

Commit:
dzg_loramod-v2:
* only return the latest value as a single reading

Commit:
dzg_loramod: Make it clear to not use the previous version.

Commit:
netop:
* added new parser to handle all netop devices. Currently both door sensors and the manhole sensor are included

Commit:
netop:
* correct order of connected slots

h4cc committed:
2019-05-13 [jb]: Added interpolation feature. Added registers with full OBIS codes.

Commit:
netop:
* added missing documentation

h4cc committed:
2019-05-14 [jb]: Added full obis code if available.

Commit:
nas_um30x3:
* added testcases for mbus parsing

Commit:
Updated wmbus lib to latest commit

Mr-SvomF committed:
Create lobaro_pressure26d.ex

Mr-SvomF committed:
Update README.md

h4cc committed:
2019-05-14 [jb]: Added full obis code if available. Added interpolation feature.

Commit:
Update libwmbus to newest commit.

Commit:
nas_luminaire:
* remove warning

h4cc committed:
2019-05-14 [jb]: Rounding all values as float to a precision of 3 decimals.

Commit:
added yabby GPS Parser

Mr-SvomF committed:
Merge pull request #14 from kleinernik/patch-4

add state field back for map visualisation

h4cc committed:
2019-05-15 [jb]: Rounding all values as float to a precision of 3 decimals.

Commit:
update klax

Commit:
updated yabby tests function

Commit:
fixed yabby tests

Commit:
Merge branch 'master' of github.com:ZennerIoT/element-parsers

Commit:
Merge branch 'master' into tabs_all_in_one

# Conflicts:
#	README.md

Commit:
tabs: Adjust according to changes by NK in tabs_doornwindow.ex

h4cc committed:
Merge pull request #19 from ZennerIoT/tabs_all_in_one

Tabs all in one

h4cc committed:
2019-05-16 [jb]: Removed/changed fields (meter_id, manufacturer_id, state). Added interpolation feature. Added obis codes.

h4cc committed:
Raising elixir version to 1.8 from 1.7

h4cc committed:
GWF Meter documentation and naming fixed.

Commit:
* update wmbus lib, which now uses atoms in keys
* fixed tests using the wmbus lib

h4cc committed:
2019-06-17 [jb]: Added obis field for gas_in_liter. Added interpolation of values. Fixed boot message for v0.7.0.

h4cc committed:
Fixed broken test.

Commit:
nas_um30x3:
* shutdown message now returns optional status message as well
* boot message contains firmware version

h4cc committed:
Fixed profile access example.

h4cc committed:
Update example_profile_read.ex

Commit:
elvaco_cma11l:
* added new parser for CMa11L by elvaco

h4cc committed:
Clarified what happend with the tabs parsers.

h4cc committed:
Fixed tests, seems to be a problem with float numbers and rounding.

Mr-SvomF committed:
Delete lobaro_wmbus-bridge.ex

Outdated, no longer needed since ELEMENT now integrates wMBus-Bridge by default

h4cc committed:
2019-06-20 [jb]: Added medium "heatcostallocator".

h4cc committed:
2019-07-11 [jb]: Added handling of a-plus-a-minus with register2_value field.

h4cc committed:
2019-07-11 [jb]: Added handling of a-plus-a-minus with register2_value field. ELEMENT compatability.

h4cc committed:
2019-07-11 [jb]: Added handling of a-plus-a-minus with register2_value field. ELEMENT compatability. Fixed Typo.

h4cc committed:
2019-06-21 [jb]: Added handling of a-plus-a-minus with register2_value field.

eagle-te committed:
* added parse for GWF meter type 0x02
* added doc with possible relevant information, regarding meter_id BCD parsing
* fixed spelling of continuous
* added tests for type 0x02

Signed-off-by: eagle-te <breitenstein.alexander@gmail.com>

eagle-te committed:
* fixed missing changelog
* added additional documentation

Signed-off-by: eagle-te <breitenstein.alexander@gmail.com>

Commit:
Update lib_wmbus to most recent version (mode 7 encryption + long header)

eagle-te committed:
* parser for DL-TRS12 added
* added test and cleaned up code

Signed-off-by: eagle-te <breitenstein.alexander@gmail.com>

Commit:
Update lib_wmbus to most recent version (bugfix for mode 5 with long header; vib with volume, unknown vife and variable length)

Commit:
Update lib_wmbus to most recent version (implemented mode 0)

Commit:
Update lib_wmbus to most recent version (add information to message)

h4cc committed:
2019-07-15 [jb]: Fixing interpolation for changes between measurements in a packet.

Commit:
Update lib_wmbus to most recent version (add din address to message content)

Mr-SvomF committed:
Create elvaco_cmi4110.ex

Mr-SvomF committed:
Update README.md

Mr-SvomF committed:
Update elvaco_cmi4110.ex

added Logger

Commit:
Update lib_wmbus to most recent version (fix wrong output of din_address)

Commit:
Merge branch 'master' of github.com:ZennerIoT/element-parsers

h4cc committed:
2019-07-23 [jb]: Skipping negative values by default.

h4cc committed:
2019-07-31 [jb]: Added :obis_value key for first register value, supporting MSCONS rule.

h4cc committed:
Fixed tests.

h4cc committed:
Added lib/nas_cm3060.ex

h4cc committed:
Added nas_cm3060 fields()

h4cc committed:
Fixed reset_reason() in nas_3060.ex

Commit:
Updated NAS CM3010 to have a catchall parse function and changelog.
Also removed rssi from status message.

h4cc committed:
2019-08-28 [jb]: Added missing :obis_value key in message format v2.

h4cc committed:
Fixed tests.

h4cc committed:
Added NAS CM3030

h4cc committed:
Fixed readme.

h4cc committed:
Merge pull request #21 from eagle-te/decentlab-dl-trs12-parser

Parser for DL-TRS12

h4cc committed:
Merge pull request #20 from eagle-te/type-0x02-gwf-watermeter-parser

GWF meter type 0x02 support

h4cc committed:
Fixed GWF Parser.

h4cc committed:
Fixed README plus tests.

h4cc committed:
Added dummy parser generating GPS coordinates.

h4cc committed:
Removed Logger.info(error)

h4cc committed:
Added changelog and parsing catchall to adeunis parsers.

h4cc committed:
Added changelog and parsing catchall to adeunis parsers. Fixup.

h4cc committed:
Added changelog and parsing catchall to ascoel, bosch and comtac parsers.

h4cc committed:
Added changelog and parsing catchall to rest of parsers.

h4cc committed:
Removed lobaro wmbus bridge link.

fl1x0r committed:
added parser for globalsat and netvox

h4cc committed:
Merge pull request #24 from fl1x0r/master

added parser for globalsat and netvox

h4cc committed:
Fixed tests for new parsers.

h4cc committed:
Added sensative strips parser.

h4cc committed:
2019-09-19 [jb]: Ignoring invalid payloads. Handling missing GPS fix.

h4cc committed:
Echoing name of test.

h4cc committed:
2019-09-30 [jb]: Using meta.transceived_at instead of DateTime.utc_now in add_power_from_last_reading()

h4cc committed:
2019-09-30 [jb]: Initial implementation

h4cc committed:
Added lib/example_device_location.ex to readme

h4cc committed:
Fixed helper function get/3 to work with tuples, like device.location.coordinates.

h4cc committed:
Added parser lobaro_oskar_v2.ex

h4cc committed:
2019-10-10 [jb]: Added key interpolated=1 and :obis_value to interpolated readings.

h4cc committed:
2019-10-15 [jb]: Initial implementation according to "T0005370_TRM_ver2.0.pdf"

h4cc committed:
2019-10-15 [jb]: Initial implementation according to "T0005370_TRM_ver2.0.pdf"
Fixed external input to 1 instead of 255

h4cc committed:
2019-10-15 [jb]: Initial implementation according to "T0005370_TRM_ver2.0.pdf"
Inverted external_input and reed_switch values.

h4cc committed:
2019-10-16 [jb]: Initial implementation according to "sensoneo_quatro_QS_Payload_v2.pdf".

h4cc committed:
2019-10-30 [jb]: Field server_id is now integer. Added hex string value server_id_hex.

h4cc committed:
Fixed parsing catchall.

h4cc committed:
Testing all parsers for a working catchall.

h4cc committed:
Added NAS CM3020 Parser.

h4cc committed:
Fixes.

h4cc committed:
Added NAS CM3020 to README.

h4cc committed:
Added tests to lobaro oskar v2

h4cc committed:
2019-11-05 [jb]: Fixed order of temperature/humidity in 1.1.17, 1.1.18 and 1.1.19.

h4cc committed:
The first field should be a numeric value, so it can be used for graphs.

h4cc committed:
2019-11-15 [jb]: Handling 32bit usage payload too.

h4cc committed:
Added NKE remote temperature parser.

Mr-SvomF committed:
Update lobaro_pressure26d.ex

h4cc committed:
2019-11-29 [jb]: Removed all reverse engineered fields because they changed.

h4cc committed:
Added Cayenne LPP Parser.

h4cc committed:
Fixed naming to .ex
Added check in test.sh for that.

h4cc committed:
Added more tests.

h4cc committed:
2019-12-09 [jb]: Fixed boot message for newer longer payloads.

h4cc committed:
2019-12-10: [jb] Fixed mapping in reset_cause/1

h4cc committed:
2019-12-10: [jb] Fixed mapping in reset_cause/1

h4cc committed:
2019-12-10 [jb]: Initial implementation according to "PNI PlacePod Sensor - Communications Protocol.pdf"

h4cc committed:
2019-12-11 [jb]: Added field obis_value in all readings for MSCONS rule compatability.

h4cc committed:
2019-12-12 [jb]: Ignoring frame_port for messages without frame header, fame_port=8 was required before.

h4cc committed:
2019-06-05: [gw] refactoring. Checked with v1.8.5

h4cc committed:
2019-12-27 [jb]: Added field "Proximity in %"

h4cc committed:
2020-01-08 [jb]: Added Taupunkttemperatur calculation.

h4cc committed:
2020-01-08 [jb]: Added fields for Indoor Localization

Commit:
added some devices to netvox multiparser

Mr-SvomF committed:
Update netvox_multiparser.ex

Mr-SvomF committed:
Update netvox_multiparser.ex

Commit:
Added Payload Version 5.0

Mr-SvomF committed:
Create lobaro_environmental.ex

h4cc committed:
2020-01-27 [jb]: Added taupunkt calculation

Commit:
Corrected tests for Lobro GPS Tracker payload

Commit:
Merge branch 'master' of https://github.com/ZennerIoT/element-parsers

h4cc committed:
Added parser lib/lobaro_environmental.ex to README.md

h4cc committed:
2020-01-27: [jb] Fixed handling of tank.form profile field.

h4cc committed:
Added EBZ electricity meter parser.

h4cc committed:
Updated lib_wmbus dependency.

h4cc committed:
2020-02-04 [jb]: Initial implementation according to "Axioma Lora Payload W1 V01.7 Extended.pdf"

h4cc committed:
Fixes

Mr-SvomF committed:
added current meter

Commit:
[as] Added RFI Remote Switch Parser

Mr-SvomF committed:
Update rfi_remote-power-switch.ex

Mr-SvomF committed:
Update rfi_remote-power-switch.ex

Mr-SvomF committed:
Update rfi_remote-power-switch.ex

h4cc committed:
2020-03-31/jb: Allowing all device_ids now

h4cc committed:
2020-04-07 [jb]: Added all missing sTypes. Fixed negative temperature bugs. Removed offset=0 values.

h4cc committed:
Added easymeter_esys_lr10.ex
Added nas_cm3061.ex

h4cc committed:
Updated parsers.

h4cc committed:
Fixed typo.

Mr-SvomF committed:
Added 2-Gang Water Leak Detector

h4cc committed:
2020-05-04 [jb]: Added config flag "config_add_last_real_time_data_to_button_press?"

h4cc committed:
Merge remote-tracking branch 'origin/master'

h4cc committed:
2020-05-04 [jb]: Added config flag "config_add_last_real_time_data_to_button_press?"

h4cc committed:
Fixed example DOWN messages.

h4cc committed:
2020-05-12 [jb]: Fixed single temperature value.

h4cc committed:
2020-05-13 [jb]: Fixed interpolate: false. Made configuration testable.

h4cc committed:
Added docs.

h4cc committed:
2020-05-27 [nk]: Initial implementation of new version

h4cc committed:
Added Sontext parser
2020-06-09 [jb]: Initial implementation according to "M-Bus Frames 7x9 - LoRAWAN_20190812.pdf"

h4cc committed:
Added Tektelic Industrial GPS Asset Tracker
2020-06-11 [jb]: Initial implementation according to "T0006279_TRM_ver0.6.r1.pdf"

h4cc committed:
2020-06-12 [jb]: Supporting v4 Payloads, RENAMED READING FIELDS!

h4cc committed:
2020-06-16 [jb]: Fixed call to add_location. Always creating locations ignoring gps_valid flag.

Mr-SvomF committed:
added 3 phase current meter

h4cc committed:
2020-06-18 [jb]: Initial implementation according to "ATIM_ACW-DIND160-80-88-44_UG_EN_V1.7.pdf"

h4cc committed:
2020-06-23 [jb]: Added feature :append_last_digital_inputs default: false

h4cc committed:
2020-06-26 [jb]: Using memory_address, sub_device and tariff in reading keys.

h4cc committed:
Updating wmbus parsers.

h4cc committed:
Added abeeway industrial tracker.

h4cc committed:
Added axioma_qalcosonic.ex

h4cc committed:
2020-07-02 [jb]: Updated payload to Strips-MsLoRa-DataFrames-3.odt, ignoring frame_port=2 for now.

h4cc committed:
Added parametric_pcr2.ex

h4cc committed:
Fixed fields.

h4cc committed:
Added parametric_tcr.ex

h4cc committed:
2020-07-08 [jb]: Added all error flags in build_error_string()

narrowtux committed:
Ignore sensoneo

h4cc committed:
Sensoneo Parsers on request.

h4cc committed:
Fixing parser whitelist problem.

h4cc committed:
2020-07-23 [jb]: Initial implementation according to "Payload spec command V203.pdf"

h4cc committed:
Fixed xter connect parser naming.

h4cc committed:
Added nke_smart_plug.ex

h4cc committed:
2020-08-26 [jb]: Support for v3 payloads.

h4cc committed:
2020-08-27 [jb]: Added alerts for v3 payloads.

h4cc committed:
Fixed typo.

Commit:
Added support for new Healthy Home Sensor version (with indoor-air-quality)

Commit:
Healthy home sensor only uses half of the byte for battery state and (supposedly only has 14 steps of battery state opposed to previous 15 steps)

Commit:
Healthy home sensor only uses half of the byte for battery state and (supposedly only has 14 steps of battery state opposed to previous 15 steps)

h4cc committed:
2020-09-29 [jb]: Initial implementation according to https://parametric.ch/docs/tcr/tcr_payload_v2

h4cc committed:
Updated comment.

flixbi committed:
Parser for the Agriculture Sensor

👆 That's the one you stick into mother earth 🌍🌱

h4cc commented on lib/agriculture_sensor_parser.ex:
```diff
@@ -0,0 +1,354 @@
+          watermark_02: 0,
+          ambient_light_intensity: 0,
+          ambient_light_alarm: false,
+          accelerometer_data: %{x: 0, y: 0, z: 0},
```
No sub-structures possible. Use `accelerometer_x` ... instead.

flixbi replied:
Ok, I hoped it would at least render to an JSON-Array. Is there a reason for the use of a plain data format?

h4cc commented on lib/agriculture_sensor_parser.ex:
```diff
@@ -0,0 +1,354 @@
+        [parsed_payload]
+        
+      # Default behaviour -> no bytes left
+      _ -> 
```
Can only be `<<>>`, the empty binary.

flixbi replied:
The underscore is the placeholder for any value, equal to "default" in other languages, isn't it? 
Or did you ment the relevance for the upper part `<< x::binary >> when byte_size(x) > 0 ->`?
It's an possibility, that in future firmware releases of the agriculture sensor the Payload might extend,
for this case I print an error with the rest of the Payload itself as a part of the returned error. 
I thought it would be an greater solution than just silent fail or just log it (when now one is watching logs).

If you ment the default behaviour, the empty Bitstring `<<>>` is also possible, but in the style of the case element, I would preffer the Underscore `_`, for this case I'll switch it to the Bitstring. 💪

h4cc commented on lib/agriculture_sensor_parser.ex:
```diff
@@ -0,0 +1,354 @@
+
+      # battery_remaining_lifetime | 1 Byte | Fakt: 1%/LSB
+      << 0x00, 0xFF, battery_remaining_lifetime::8 , overhead::binary >> ->  
+        parse_dyn(
```
Using a pipe would look cool:

```
parsed_payload
|> Map.put_new
|> parse_dyn
```

flixbi replied:
I'll change it for readability. 👍

h4cc commented:
Here are two more test payloads, could you please integrate them?

```
0104057a020202f00965002b0b6700d50b685c
0104057a020202d3096500220b6700e20b685b
0104056d02020446096503b10b67005a0b68aa
```

flixbi committed:
Update agriculture_sensor_parser.ex

- Style changes (using pipes) in recursive "parse_dyn" function, which makes it a bit slimmer. 
- changed handling of empty bitstrings in the default behaviour of case statement in the "parse_dyn" function.
- optimized map in "accelerometer_data" and divided it up to a single value per Achsis so: "accelerometer_{x,y,z}"
- added more test-payloads.
